### PR TITLE
[PP-8] Add single service to make requests to be

### PR DESCRIPTION
**Pull request by andreikaptsiuh**
Task: https://andrei-lysiuk.atlassian.net/browse/PP-8
Description: Create class that will make requests to BE via axios lib

### CODE REVIEW

**Review by mrakofff — CHANGES_REQUESTED** [review]
Resolve conflicts with `develop`

File: src/api/axios.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { ICard, IRoom, ITask, IUser } from '@/interfaces';
```
Comment (mrakofff) [inline src/api/axios.ts]: 👍

File: src/api/axios.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class Axios {
+	private static classInstance?: Axios;
+
+	readonly instance: AxiosInstance;
```
Comment (mrakofff) [inline src/api/axios.ts]: Should it be `private`?

File: src/api/axios.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  constructor(baseURL: string){
```
Comment (mrakofff) [inline src/api/axios.ts]: Could be `private` as well. BaseURL should be `optional` coz we will have separate class for BE links

File: src/api/axios.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import { ICard, IRoom, ITask, IUser } from '@/interfaces';
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class Axios {
```
Comment (mrakofff) [inline src/api/axios.ts]: It's better to set it like `Http` and `this.instance` it should be an instance of this class, but not an instance of `axios`. It's better to use `_method` to show that `method/variable` is private. So you could have `_instance: Http` and `_http: AxiosInstance` fields.

File: src/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+export interface ITask {
+  value: string;
+  id?: number;
```
Comment (mrakofff) [inline src/interfaces.ts]: All `ids` most of the time will be a `string` type

File: src/api/axios.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    console.log('Config:', error.config);
+  };
+
+  getUsers = () => this.instance.get<IUser[]>('/users');
+  getUserById = (id: number) => this.instance.get<IUser>(`/users/${id}`);
+  createUser = (body: IUser) => this.instance.post('/users', body);
+  deleteUser = (id: number) => this.instance.delete(`/users/${id}`);
+
+  getTasks = () => this.instance.get<ITask[]>('/tasks');
+  getTaskById = (id: number) => this.instance.get<ITask>(`/tasks/${id}`);
```
Comment (mrakofff) [inline src/api/axios.ts]: use generic type of `ITask`

File: src/api/axios.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  private _initializeResponseInterceptor = () => {
+    this.instance.interceptors.response.use(
+      this._handleResponse,
+      this._handleError,
+    );
+  };
+
+  private _handleResponse = ({ data }: AxiosResponse) => data;
+  
+  protected _handleError = (error: AxiosError) => {
```
Comment (mrakofff) [inline src/api/axios.ts]: why it'is `protected`?

File: src/api/axios.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    if (error.response) {
```
Comment (mrakofff) [inline src/api/axios.ts]: I think it's redundant and you could log only `Error` object, if you need you could print it through the `console.dir`

File: src/api/axios.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  getUsers = () => this.instance.get<IUser[]>('/users');
+  getUserById = (id: number) => this.instance.get<IUser>(`/users/${id}`);
+  createUser = (body: IUser) => this.instance.post('/users', body);
+  deleteUser = (id: number) => this.instance.delete(`/users/${id}`);
+
+  getTasks = () => this.instance.get<ITask[]>('/tasks');
+  getTaskById = (id: number) => this.instance.get<ITask>(`/tasks/${id}`);
+  createTask = (body: ITask) => this.instance.post('/tasks', body);
+  deleteTask = (id: number) => this.instance.delete(`/tasks/${id}`);
```
Comment (mrakofff) [inline src/api/axios.ts]: I think it's out of scope of this task. Also BE requests will be in a separate `class`

**Review by mrakofff — CHANGES_REQUESTED** [review]
1. Add comments with naming, additional code. 
1. Good point with interface folder 👍 
1. Add basic `tests`

File: src/api/axios.ts (hunk 2 above)
Comment (andreikaptsiuh) [reply src/api/axios.ts]: fixed

File: src/api/axios.ts (hunk 3 above)
Comment (andreikaptsiuh) [reply src/api/axios.ts]: How i can make it? I not should add create instance?

File: src/api/axios.ts (hunk 9 above)
Comment (andreikaptsiuh) [reply src/api/axios.ts]: deleted

File: src/api/axios.ts (hunk 4 above)
Comment (andreikaptsiuh) [reply src/api/axios.ts]: complete

File: src/api/axios.ts (hunk 7 above)
Comment (andreikaptsiuh) [reply src/api/axios.ts]: fixed

File: src/interfaces.ts (hunk 5 above)
Comment (andreikaptsiuh) [reply src/interfaces.ts]: fixed

File: src/api/axios.ts (hunk 3 above)
Comment (mrakofff) [reply src/api/axios.ts]: https://metanit.com/web/typescript/3.4.php

File: src/api/HttpClient.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class HTTPClient {
+	static classInstance?: HTTPClient;
+
+	private instance!: AxiosInstance;
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: rename to `_http` || `_httpClient` and why do you need `!` symbol?

File: src/api/HttpClient.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class HTTPClient {
+	static classInstance?: HTTPClient;
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: rename to `_instance` and why it's `optional`?

**Review by mrakofff — CHANGES_REQUESTED** [review]
Did you test it? does it works? coz i have some doubts with `static _instance`. If it does so good job 👍 
Do not forget to create some basic tests

File: src/api/HttpClient.ts (hunk 10 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/api/HttpClient.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+export class _HTTPClient {
+	static _instance: _HTTPClient;
+
+	private _axios!: AxiosInstance;
+
+  constructor(public baseURL = ' ') {
+    this._axios = axios.create({ baseURL: this.baseURL});
+    this._initializeResponseInterceptor();
+	}
```
Comment (mrakofff) [inline src/api/HttpClient.ts:11]: Is it problem with `github` or it's problem with formatting?

File: src/api/HttpClient.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class _HTTPClient {
+	static _instance: _HTTPClient;
+
+	private _axios!: AxiosInstance;
+
+  constructor(public baseURL = ' ') {
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: is it empty string or it's space ` `?

File: src/api/HttpClient.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
+
+export class _HTTPClient {
+	static _instance: _HTTPClient;
+
+	private _axios!: AxiosInstance;
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: Why do you need this `!` ?) Any reason? As I remember it's a sign to exclude optionality?

File: src/api/HttpClient.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+	}
+
+  static getClassInstance(){
+    if (!this._instance) {
+      this._instance = new _HTTPClient();    
+    };
+    return this._instance;
+  };
+
+  private _initializeResponseInterceptor = () => {
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: Perhaps it's better to set `_initializeInterceptors` coz we will insert request interceptor too.

File: src/api/HttpClient.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+	static _instance: _HTTPClient;
+
+	private _axios!: AxiosInstance;
+
+  constructor(public baseURL = ' ') {
+    this._axios = axios.create({ baseURL: this.baseURL});
+    this._initializeResponseInterceptor();
+	}
+
+  static getClassInstance(){
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: Could we reduce name to `getInstance` coz we have any additional instances?

**Review by mrakofff — CHANGES_REQUESTED** [review]
I hope it's latest update request:)

File: src/api/HttpClient.ts (hunk 13 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: it's space

File: src/api/HttpClient.ts (hunk 12 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts:11]: It's problem with gitHub

File: src/api/HttpClient.ts (hunk 16 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/api/HttpClient.ts (hunk 15 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: add request interceptor

File: src/api/HttpClient.ts (hunk 14 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/api/HttpClient.ts (hunk 13 above)
Comment (mrakofff) [reply src/api/HttpClient.ts]: any reason for empty space? we could have problems with empty space in requests `url`

File: src/api/HttpClient.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export class _HTTPClient {
+	static _instance: _HTTPClient;
+
+	private _axios: AxiosInstance;
+
+  constructor(public baseURL = ' ') {
+    this._axios = axios.create({ baseURL: this.baseURL});
+    this._initializeInterceptor();
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: `_initializeInterceptor` => `_initializeInterceptors`

File: src/api/HttpClient.ts (hunk 13 above)
Comment (mrakofff) [reply src/api/HttpClient.ts]: could we make constructor `private` to exclude handle creation entities?

**Review by mrakofff — CHANGES_REQUESTED** [review]
And final corrections:)

File: src/api/HttpClient.ts (hunk 17 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/api/HttpClient.ts (hunk 13 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

**Review by Praddby — APPROVED** [review]

File: src/api/HttpClient.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export class _HTTPClient {
+	static _instance: _HTTPClient;
+
+	_axios: AxiosInstance;
+
+  constructor(private baseURL = '') {
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: update with `private constructor`

File: src/api/HttpClient.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export class _HTTPClient {
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: remove `_` from class name

File: src/test/HttpClient.test.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+    expect(httpClient).not.toBe(undefined);
+  });
+  
+  test('Should be one class', () => {
+    const httpClientDouble = _HTTPClient.getInstance();
+    expect(httpClient === httpClientDouble).toBe(true);
+  });
+});
+
+describe('Response', () => {
```
Comment (mrakofff) [inline src/test/HttpClient.test.ts]: You don't need this test, add test for `baseUrl`. I think it will be enough

File: src/api/HttpClient.ts [hunk 21]
```diff
@@ -0,0 +1,39 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+
```
Comment (mrakofff) [inline src/api/HttpClient.ts:2]: Add please import of `http` or `httpClient` variable that will be equal to `HttpClient.getInstance().http`

File: src/api/HttpClient.ts [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  private _handleResponse = ({ data }: AxiosResponse) => data;
+
+  private _handleRequest = (config: AxiosRequestConfig) => config;
+  
+  private _handleError = (error: AxiosError) => {
+    console.dir(error);
+  };
+};
```
Comment (mrakofff) [inline src/api/HttpClient.ts:47]: Please add `http` getter that will return `this._axios`

**Review by mrakofff — CHANGES_REQUESTED** [review]

File: src/api/HttpClient.ts (hunk 19 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/api/HttpClient.ts (hunk 21 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts:2]: add

File: src/api/HttpClient.ts (hunk 22 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts:47]: complete

File: src/test/HttpClient.test.ts (hunk 20 above)
Comment (andreikaptsiuh) [reply src/test/HttpClient.test.ts]: fix

File: src/api/HttpClient.ts (hunk 22 above)
Comment (mrakofff) [reply src/api/HttpClient.ts:47]: it should not be an `axios` name. should be some abstract. 
About `getters` https://www.typescripttutorial.net/typescript-tutorial/typescript-getters-setters/

File: src/test/HttpClient.test.ts [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  test('Should return clean string', () => {
+    const axiosBaseUrl = axios.defaults.baseURL;
+    expect(axiosBaseUrl === '').toBe(true);
+  });
+
+  test('Should return new url', () => {
+    const url = 'new url';
+    axios.defaults.baseURL = url;
```
Comment (mrakofff) [inline src/test/HttpClient.test.ts]: we should prevent overriding private `HttpClient` fields

File: src/api/HttpClient.ts [hunk 24]
```diff
@@ -0,0 +1,43 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export class HTTPClient {
+	static _instance: HTTPClient;
+
+	_axios: AxiosInstance;
```
Comment (mrakofff) [inline src/api/HttpClient.ts]: make it private. we should hide request realization from outside

File: src/test/HttpClient.test.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+});
+
+describe('Base url', () => {
+  const axios = httpClient.getAxios();
+
+  test('Should return axios', () => {
+    expect(axios).not.toBe(undefined);
+  });
+
+  test('Should return clean string', () => {
```
Comment (mrakofff) [inline src/test/HttpClient.test.ts]: You should test that call requester will be with correct arguments, and passing baseUrl to constructor will affect any request after

**Review by mrakofff — CHANGES_REQUESTED** [review]

File: src/test/HttpClient.test.ts (hunk 20 above)
Comment (andreikaptsiuh) [reply src/test/HttpClient.test.ts]: fix

File: src/api/HttpClient.ts (hunk 18 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: update

File: src/api/HttpClient.ts (hunk 24 above)
Comment (andreikaptsiuh) [reply src/api/HttpClient.ts]: fix

File: src/test/HttpClient.test.ts (hunk 25 above)
Comment (andreikaptsiuh) [reply src/test/HttpClient.test.ts]: fix

**Review by mrakofff — APPROVED** [review]

**Pull request merged**